### ISSUE TITLE
Vec-of-array

### ISSUE THREAD

**Issue by Kromey** [Kromey/serde_arrays#2]
While `Vec` is `Serialize`/`Deserialize`, const generic and large arrays are not, meaning that `Vec`s containing these can not be handled through Serde. Neither does this library (currently) handle them, since we're unable to apply Serde's traits to these types due to Rust's orphan rule.

We should however be able to implement support for `Vec<[T; N]>`, as well as `[[T; N]; M]`, which I expect should cover most use cases where these structures are being nested; one example of the former can be seen [here](https://github.com/sdd/kiddo/blob/master/src/custom_serde.rs).

Unfortunately, it appears that without support from Serde itself, supporting arbitrary levels of nesting is not possible.

**Comment by jonasbb** [comment]
> Unfortunately, it appears that without support from Serde itself, supporting arbitrary levels of nesting is not possible.

That is not true as you can [see here](https://docs.rs/serde_with/1.8.1/serde_with/guide/serde_as/index.html#big-array-support-rust-151). You can arbitrarily nest types here. Both array-in-vec and array-in-array are supported, but also array-in-hashmap, etc. The crate is implemented without any support from serde.

**Comment by Kromey** [comment]
Thank you for that clarification and correction! I'm going to have to look deeper into how `serde_with` does this, at a quick glance they seem to be doing exactly what I was trying to do and running into insurmountable compiler errors due to "conflicting implementations". Clearly they did something different than I tried to do, it will just take some time to dig into it and figure out what.

It seems a great crate (that I had no idea existed until after I published this one), though I really don't like the syntax they use, so if I can find a "cleaner" way to approach it I think my crate can still have something to offer.

**Comment by jonasbb** [comment]
I can give you some hints where to look. `serde_with` works with two traits `SerializeAs` and `DeserializeAs`. The function signatures are mostly compatible with `Serialize`/`Deserialize`. There are two types `As` and `Same` which act like adapters to glue the two kinds of traits together. This gluing together is also what the `#[serde_as]` attribute on the struct does.
You need a `SerializeAs` implementation for each container type (e.g., array, `Vec`, `BTreeMap`, available in `src/ser/*`). That is analogue to the `Serialize implementation in `serde`.

The `SerializeAs` trait is defined "backwards" which makes it in some ways more flexible than `Serialize`. You can easily add now conversion types, and they will compose well with each other.

The `SerializeAs` trait, was invented by markazmierczak [over here](https://github.com/serde-rs/serde/issues/723#issuecomment-565799505). I adapted and extended them for `serde_with` though.

**Comment by brainstorm** [comment]
Oh, no, I really needed this for my current toy project :_/

```
/// N encodes password, only 65535 login/password combinations possible
#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
struct LoginPasswordBruteForcer<const N: usize> {
    #[serde(with = "serde_arrays")]
    state: [State; N],
    #[serde(with = "serde_arrays")]
    good_keys: [u16; N],
    #[serde(with = "serde_arrays")]
    bad_keys: [[u16; N]; N],
}
```

I wanted to serialize/deserialize the login/password attempt "sessions" so that I don't have to start from scratch when re-starting my program, hence `bad_keys` would store the failed (permission denied) attempts. 

The discussion (and work) being done here is a bit over my current Rust skill level, so I'm just asking as a user (kudos for this crate, btw!): will nested Serialization/Deserialization be available anytime soon or `Deserialize` is still proving elusive, @Kromey?

**Comment by Kromey** [comment]
@brainstorm I had some personal/family issues come up that suffice to say distracted me from even thinking about this. It has crept back onto my radar recently though, and I'm hoping to tackle this soon. The trick isn't so much implementing nested arrays, but doing so in a manner that won't leave me coded into a corner should I want to expand that export further later.

**Comment by Kromey** [comment]
Partial support (serialization of e.g. `Vec<[T; N]>`) is now available in v0.2.0.

I'm leaving this issue open because I still want to support deserialization as well; currently that requires a wrapper struct, or the `serde_with` crate can do it.
